<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откроем-и-изучим-файл" data-toc-modified-id="Откроем-и-изучим-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откроем и изучим файл</a></span><ul class="toc-item"><li><span><a href="#Импортируем-необходимые-для-работы-модули-и-библиотеки" data-toc-modified-id="Импортируем-необходимые-для-работы-модули-и-библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импортируем необходимые для работы модули и библиотеки</a></span></li><li><span><a href="#Открываем-таблицу" data-toc-modified-id="Открываем-таблицу-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открываем таблицу</a></span></li><li><span><a href="#Просмотрим-основные-данные-о-датасете" data-toc-modified-id="Просмотрим-основные-данные-о-датасете-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Просмотрим основные данные о датасете</a></span></li></ul></li><li><span><a href="#Разобьём-данные-на-выборки" data-toc-modified-id="Разобьём-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разобьём данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Обозначим-признаки-как-features,-а-целевой-признак---как-target" data-toc-modified-id="Обозначим-признаки-как-features,-а-целевой-признак---как-target-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обозначим признаки как features, а целевой признак - как target</a></span></li><li><span><a href="#Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки.-Примем-соотношение-60/20/20%-соответственно" data-toc-modified-id="Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки.-Примем-соотношение-60/20/20%-соответственно-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Разделим исходные данные на обучающую, валидационную и тестовую выборки. Примем соотношение 60/20/20% соответственно</a></span></li></ul></li><li><span><a href="#Исследуем-модели" data-toc-modified-id="Исследуем-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуем модели</a></span><ul class="toc-item"><li><span><a href="#Используем-дерево-решений" data-toc-modified-id="Используем-дерево-решений-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Используем дерево решений</a></span></li><li><span><a href="#Попробуем-случайный-лес" data-toc-modified-id="Попробуем-случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Попробуем случайный лес</a></span></li><li><span><a href="#Теперь-проверим-логистическую-регрессию" data-toc-modified-id="Теперь-проверим-логистическую-регрессию-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Теперь проверим логистическую регрессию</a></span></li></ul></li><li><span><a href="#Проверим-модель-на-тестовой-выборке" data-toc-modified-id="Проверим-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверим модель на тестовой выборке</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#(бонус)-Проверим-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверим-модели-на-адекватность-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>(бонус) Проверим модели на адекватность</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Рекомендация тарифов

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».

Перед нами стоит задача классификации.



В нашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — мы её уже сделали.

Построим модель с максимально большим значением *accuracy*. Проверим *accuracy* на тестовой выборке.

Откроем файл с данными и изучим его.

Разделим исходные данные на обучающую, валидационную и тестовую выборки.

Исследуем качество разных моделей, меняя гиперпараметры. Кратко напишем выводы исследования.

Проверим качество модели на тестовой выборке.


Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:

сalls — количество звонков,

minutes — суммарная длительность звонков в минутах,

messages — количество sms-сообщений,

mb_used — израсходованный интернет-трафик в Мб,

is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Откроем и изучим файл

### Импортируем необходимые для работы модули и библиотеки

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from joblib import dump
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score

### Открываем таблицу

In [2]:
try:
    data = pd.read_csv('/datasets/users_behavior.csv', sep=',')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv', sep=',')

In [3]:
data.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:

сalls — количество звонков,

minutes — суммарная длительность звонков в минутах,

messages — количество sms-сообщений,

mb_used — израсходованный интернет-трафик в Мб,

is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

### Просмотрим основные данные о датасете

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
data.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [6]:
data.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [7]:
data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [8]:
#data['is_ultra'].value_counts()
for col in data.columns:
    print(data[col].value_counts())
    print('Среднее значение:', data[col].mean())
    print()

69.0     60
68.0     49
64.0     48
65.0     47
77.0     47
         ..
161.0     1
189.0     1
162.0     1
132.0     1
244.0     1
Name: calls, Length: 184, dtype: int64
Среднее значение: 63.03889234598631

0.00      40
2.00       3
1.00       3
564.01     3
561.37     2
          ..
135.52     1
618.87     1
395.23     1
542.54     1
722.64     1
Name: minutes, Length: 3144, dtype: int64
Среднее значение: 438.20878655880523

0.0      497
20.0      52
6.0       46
16.0      45
17.0      44
        ... 
138.0      1
176.0      1
125.0      1
150.0      1
190.0      1
Name: messages, Length: 180, dtype: int64
Среднее значение: 38.28126944617299

0.00        11
0.01         2
27210.40     1
18932.66     1
12762.99     1
            ..
8152.11      1
10204.48     1
5426.77      1
22924.34     1
15613.81     1
Name: mb_used, Length: 3203, dtype: int64
Среднее значение: 17207.673836341008

0    2229
1     985
Name: is_ultra, dtype: int64
Среднее значение: 0.30647168637212197



Как видим, датасет хороший, нет нулевых значений, нет дубликатов. У 40 человек количество минут равно нулю, у 497 число сообщений, у 11 человек количество использованных мегабайт. 30% людей пользовались тарифом "Ультра", остальные - нет.

## Разобьём данные на выборки

### Обозначим признаки как features, а целевой признак - как target

In [9]:
features = data.drop(['is_ultra'], axis=1)
target = data['is_ultra']

### Разделим исходные данные на обучающую, валидационную и тестовую выборки. Примем соотношение 60/20/20% соответственно

In [10]:
features_train, features_valid, target_train, target_valid = (
    train_test_split(
        features, target, test_size=0.4,
        random_state=12345, stratify = target))

Получили тренировочную выборку = 60%, валидационную - 40%. Теперь разобьём валидационную пополам, и получим валидационную и тестовую

In [11]:
features_valid, features_test, target_valid, target_test =(
    train_test_split(
        features_valid, target_valid, test_size=0.5,
        random_state=12345, stratify=target_valid))

In [34]:
for sample in [features_train, features_valid, features_test, target_train, target_valid, target_test]:
    print(f'Размер выборки: {sample.shape}')

Размер выборки: (1928, 4)
Размер выборки: (643, 4)
Размер выборки: (643, 4)
Размер выборки: (1928,)
Размер выборки: (643,)
Размер выборки: (643,)


In [21]:
features_train.shape

(1928, 4)

In [22]:
features_valid.shape

(643, 4)

In [23]:
features_test.shape

(1928,)

In [24]:
target_train.shape

(643,)

In [25]:
target_valid.shape

(643, 4)

In [26]:
target_test.shape

(643,)

## Исследуем модели

У нас задача классификации, значит подходят только эти модели: дерево решений, случайный лес и логическая регрессия

Обучим модель на обучающей выборке, оценим её качество и подберём гиперпараметры на валидационной выборке

### Используем дерево решений

In [38]:
best_model = None
best_result = 0
best_depth = 0
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train) # обучим модель на тренировочной выборке
    predictions_valid = model.predict(features_valid) # получим предсказания модели на валидационной выборке
    result = accuracy_score(target_valid, predictions_valid) # посчитаем значение метрики accuracy на валидационной выборке
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth

print("Accuracy наилучшей модели на валидационной выборке:", best_result, "Глубина дерева:", best_depth)

Accuracy наилучшей модели на валидационной выборке: 0.7853810264385692 Глубина дерева: 5


Получили accuracy 0.78, неплохо

### Попробуем случайный лес

В случае применения RandomForestClassifier при переборе гиперпараметров будем перебирать n_estimators и max_depth

In [53]:
best_model = None
best_result = 0
best_est = 0
best_depth = 0
for est in range(10, 51, 10):
    for depth in range (1, 11):
        model = RandomForestClassifier(
            random_state=12345, n_estimators=est,
            max_depth=depth, criterion='gini') # инициализируем модель RandomForestRegressor с параметрами random_state=12345, n_estimators=est и max_depth=depth
        model.fit(features_train, target_train) # обучим модель на тренировочной выборке
        predictions_valid = model.predict(features_valid) # получим предсказания модели на валидационной выборке
        result = accuracy_score(target_valid, predictions_valid) # посчитаем значение метрики accuracy на валидационной выборке
        if result > best_result:
            best_model = model
            best_result = result
            best_est = est
            best_depth = depth

print("Accuracy наилучшей модели на валидационной выборке:", best_result, "Количество деревьев:", best_est, "Максимальная глубина:", best_depth)

Accuracy наилучшей модели на валидационной выборке: 0.8211508553654744 Количество деревьев: 40 Максимальная глубина: 9


Получили accuracy 0.82, даже лучше, чем в случае с деревом решений

### Теперь проверим логистическую регрессию

In [16]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train, target_train) #обучим модель
dump(model, 'model_9_1.joblib')
predictions_valid = model.predict(features_valid) # получим предсказания модели на валидационной выборке
accuracy = accuracy_score(target_valid, predictions_valid)
accuracy

0.7387247278382582

Получили значение accuracy 0.74, менее необходимого.

Лучшие результаты - у модели Случайный лес, её и будем рассматривать далее

## Проверим модель на тестовой выборке

Наилучшую модель выбираем на валидационной выборке, у нас это случайный лес

Поэтому теперь на тестовой выборке оценим именно эту модель

In [54]:
best_model = None
best_result = 0
for est in range(10, 51, 10):
    for depth in range (1, 11):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = depth) # обучим модель с заданным количеством деревьев
        model.fit(features_train, target_train) # обучим модель на тренировочной выборке
        predictions_test = model.predict(features_test) #сделаем предсказания
        result = accuracy_score(target_test, predictions_test) # посчитаем качество модели на тестовой выборке
        if result > best_result:
            best_model = model # сохраним наилучшую модель
            best_result = result # сохраним наилучшее значение метрики accuracy на тестовых данных
            best_est = est # сохраним глубину
            best_depth = depth # сохраним количество деревьев

print("Accuracy наилучшей модели на тестовой выборке:", best_result, "Количество деревьев:", best_est, "Глубина:", best_depth)

Accuracy наилучшей модели на тестовой выборке: 0.8211508553654744 Количество деревьев: 20 Глубина: 6


Получили значение 0.82, лучше, чем необходимое минимальное 0.75

## Вывод

Целью исследования был анализ поведения клиентов и создание системы рекомендаций определённого тарифа - "Смарт" или "Ультра".

Выполнялась задача классификации. Мы построили модель, которая выберет подходящий тариф с наилучшим значением метрики *accuracy*

Были рассмотрены следующие модели:

Дерево решений, получили значение *accuracy* = 0.78 на валидационной выборке

Случайный лес, получили значение *accuracy* = 0.8211508553654744 на валидационной выборке при значении количества деревьев = 40, глубина = 9

Логистическая регрессия, получили значение *accuracy* = 0.74 на валидационной выборке


**Предсказав на тестовой выборке модели "Случайный лес", получили значение метрики accuracy = 0.8211508553654744 при гиперпараметрах n_estimators (количество деревьев) = 20, depth (глубина) = 6**
